clang/lib/Parse/ParseDecl.cpp

```diff
@@ -5559,7 +5562,6 @@ bool Parser::isTypeSpecifierQualifier() {
   case tok::kw___read_write:
   case tok::kw___write_only:
   case tok::kw___funcref:
-  case tok::kw_groupshared:
     return true;
```

Comment:
This is technically an unrelated bugfix. Not sure it's worth committing it separately, but maybe we should add a test that these keywords are not accepted in C++ mode similar to how we test some of the opencl keywords (see `clang/test/SemaOpenCLCXX/private-access-specifier.cpp`)